### ISSUE TITLE
Possible to reuse features?

### ISSUE THREAD

**Issue by timsuchanek** [ThilinaRajapakse/simpletransformers#258]
I have a binary classification task, that I want to tackle with the `ClassificationModel` and `albert` from this library.

In my case, the prediction results are not yet as great as I want them to be, so I want to label a lot more data with support of the predictions.
Concretely, I want to calculate the confidence of a result as [described here](https://www.inovex.de/blog/uncertainty-quantification-deep-learning/) and then label the samples with low confidence.

I want to achieve that by just training `n` albert models, e.g. 10 of them, with the same data.
Then I run the prediction with all of them and can see, for which predictions "they have the same opinion" (low variance), and where they have low confidence, so a high variance.

As of now, I'm naively training 10 `ClassificationModel`s, save them and then loop through all of them with `model.predict`. Then I calculate the variance using the [variance formula](https://en.wikipedia.org/wiki/Variance).

This works perfectly, however now I need 3s per data sample to calculate this.

What are some possible methods we can speed this up?
If I understand correctly, the features can probably not be reused, as all albert layers and embeddings in the models are different after the training, as all weights in albert are affected by the training.

I was thinking about using [ONNX](https://cloudblogs.microsoft.com/opensource/2020/01/21/microsoft-onnx-open-source-optimizations-transformer-inference-gpu-cpu/) to speed things up, but I'm not sure how much it'll bring and how to transform a `ClassificationModel` to ONNX.

I know I pointed to multiple things here, so let me know if I should split stuff up into multiple issues :)

**Comment by ThilinaRajapakse** [comment]
This might be a silly question but is your loop looping through the models or through the data? Also, does 3s per data sample mean it takes 3 seconds to get the prediction on a single sample with all 10 models (i.e. 0.3 seconds per sample per model)?

I am not sure how much speedup you'll get with ONNX as I haven't tried it. I did try to use it with the transformers library quite a while back but it wasn't working properly at the time so I gave it up. Unfortunately, that means I don't have any helpful info on how you might go about getting the `ClassificationModel` transformed to ONNX.

**Comment by timsuchanek** [comment]
Thanks!
Looping through the 10 models with one sample of data takes 3s in total.
So yes, the outer loop is obviously the data, but in there each of the 10 models.

**Comment by ThilinaRajapakse** [comment]
So the loop would look something like this?
```python
for batch in dataloader:
    for model in models:
```
Or is it the other way around?

**Comment by timsuchanek** [comment]
Correct, that's how the loop looks like.

**Comment by ThilinaRajapakse** [comment]
I think it would be faster if the models were in the outer loop. Did you try that? Is it possible to do it that way in your use case? That way, you can increase the eval_batch_size as only one model needs to be in GPU memory at a time.

**Comment by timsuchanek** [comment]
That's an interesting idea! I can for sure try that. I'm btw running the inference via CPU.

**Comment by timsuchanek** [comment]
For the models I get btw logs like
```
Features loaded from cache at cache_dir/cached_dev_albert_128_2_1
Features loaded from cache at cache_dir/cached_dev_albert_128_2_1
Features loaded from cache at cache_dir/cached_dev_albert_128_2_1
Features loaded from cache at cache_dir/cached_dev_albert_128_2_1
Features loaded from cache at cache_dir/cached_dev_albert_128_2_1
Features loaded from cache at cache_dir/cached_dev_albert_128_2_1
Features loaded from cache at cache_dir/cached_dev_albert_128_2_1
Features loaded from cache at cache_dir/cached_dev_albert_128_2_1
```

Does that mean, that they're able to reuse the features from the other models?

**Comment by ThilinaRajapakse** [comment]
Yes, they can (and they are being used). It can be done as long as the model and the sequence length is constant.

I'm not sure I can point you towards an exact resource but the BERT paper should contain an explanation if I'm not mistaken. Essentially, features in this context are what will be fed to a model as inputs. 

```python
class InputFeatures(object):
    """
    A single set of features of data.

        input_ids: Indices of input sequence tokens in the vocabulary.
        attention_mask: Mask to avoid performing attention on padding token indices.
            Mask values selected in ``[0, 1]``:
            Usually  ``1`` for tokens that are NOT MASKED, ``0`` for MASKED (padded) tokens.
        segment_ids: Segment token indices to indicate first and second portions of the inputs.
        label: Label corresponding to the input
    """
    def __init__(self, input_ids, input_mask, segment_ids, label_id):
        self.input_ids = input_ids
        self.input_mask = input_mask
        self.segment_ids = segment_ids
        self.label_id = label_id
```

Essentially, it contains everything the model needs to know about each sequence.

-

**Comment by ThilinaRajapakse** [comment]
Running inference on the CPU will be much slower compared to the GPU.

**Comment by timsuchanek** [comment]
Awesome, thanks for that explanation!
Ok, then I'll read the paper.

Yes, I'm aware, that CPU is much slower, that's also why I train it using a GTX 1080 ti from https://vast.ai/. 

But it's too expensive to run a GPU machine for the inference, as I would have to run it non-stop in my case.

**Comment by ThilinaRajapakse** [comment]
No problem!

Google Colab might be an option in that case.

**Comment by timsuchanek** [comment]
Yeah Google Colab, unfortunately, punishes/downgrades you when running longer jobs.

**Comment by ThilinaRajapakse** [comment]
True. :cry:
